flaky test TestFrameworkWithQuery

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
 --- FAIL: TestFrameworkWithQuery (2.06s)
    testkit.go:138: 
        	Error Trace:	pkg/testkit/testkit.go:138
        	            				pkg/testkit/testkit.go:131
        	            				pkg/disttask/framework/framework_test.go:131
        	Error:      	Received unexpected error:
        	            	[schema:1050]Table 't' already exists
        	            	github.com/pingcap/errors.AddStack
        	            		external/com_github_pingcap_errors/errors.go:174
        	            	github.com/pingcap/errors.Trace
        	            		external/com_github_pingcap_errors/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).DoDDLJob
        	            		pkg/ddl/ddl.go:1171
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).CreateTableWithInfo
        	            		pkg/ddl/ddl_api.go:2747
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).CreateTable
        	            		pkg/ddl/ddl_api.go:2586
        	            	github.com/pingcap/tidb/pkg/executor.(*DDLExec).executeCreateTable
        	            		pkg/executor/ddl.go:267
        	            	github.com/pingcap/tidb/pkg/executor.(*DDLExec).Next
        	            		pkg/executor/ddl.go:159
        	            	github.com/pingcap/tidb/pkg/executor/internal/exec.Next
        	            		pkg/executor/internal/exec/executor.go:314
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).next
        	            		pkg/executor/adapter.go:1252
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelayExecutor
        	            		pkg/executor/adapter.go:997
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelay
        	            		pkg/executor/adapter.go:823
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec
        	            		pkg/executor/adapter.go:598
        	            	github.com/pingcap/tidb/pkg/session.runStmt
        	            		pkg/session/session.go:2380
        	            	github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt
        	            		pkg/session/session.go:2239
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).ExecWithContext
        	            		pkg/testkit/testkit.go:364
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExecWithContext
        	            		pkg/testkit/testkit.go:136
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExec
        	            		pkg/testkit/testkit.go:131
        	            	pkg/disttask/framework/framework_test_test.TestFrameworkWithQuery
        	            		pkg/disttask/framework/framework_test.go:131
        	            	testing.tRunner
        	            		GOROOT/src/testing/testing.go:1595
        	            	runtime.goexit
        	            		src/runtime/asm_amd64.s:1650
        	Test:       	TestFrameworkWithQuery
        	Messages:   	sql:create table t(a int not null, b int not null), [], error stack [schema:1050]Table 't' already exists
        	            	github.com/pingcap/errors.AddStack
        	            		external/com_github_pingcap_errors/errors.go:174
        	            	github.com/pingcap/errors.Trace
        	            		external/com_github_pingcap_errors/juju_adaptor.go:15
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).DoDDLJob
        	            		pkg/ddl/ddl.go:1171
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).CreateTableWithInfo
        	            		pkg/ddl/ddl_api.go:2747
        	            	github.com/pingcap/tidb/pkg/ddl.(*ddl).CreateTable
        	            		pkg/ddl/ddl_api.go:2586
        	            	github.com/pingcap/tidb/pkg/executor.(*DDLExec).executeCreateTable
        	            		pkg/executor/ddl.go:267
        	            	github.com/pingcap/tidb/pkg/executor.(*DDLExec).Next
        	            		pkg/executor/ddl.go:159
        	            	github.com/pingcap/tidb/pkg/executor/internal/exec.Next
        	            		pkg/executor/internal/exec/executor.go:314
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).next
        	            		pkg/executor/adapter.go:1252
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelayExecutor
        	            		pkg/executor/adapter.go:997
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).handleNoDelay
        	            		pkg/executor/adapter.go:823
        	            	github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec
        	            		pkg/executor/adapter.go:598
        	            	github.com/pingcap/tidb/pkg/session.runStmt
        	            		pkg/session/session.go:2380
        	            	github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt
        	            		pkg/session/session.go:2239
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).ExecWithContext
        	            		pkg/testkit/testkit.go:364
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExecWithContext
        	            		pkg/testkit/testkit.go:136
        	            	github.com/pingcap/tidb/pkg/testkit.(*TestKit).MustExec
        	            		pkg/testkit/testkit.go:131
        	            	pkg/disttask/framework/framework_test_test.TestFrameworkWithQuery
        	            		pkg/disttask/framework/framework_test.go:131
        	            	testing.tRunner
        	            		GOROOT/src/testing/testing.go:1595
        	            	runtime.goexit
        	            		src/runtime/asm_amd64.s:1650
FAIL
================================================================================
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

TestFrameworkWithQuery has a panic 

```
panic: Log in goroutine after TestFrameworkWithQuery has completed: 
	Error Trace:	pkg/disttask/framework/framework_test.go:44
	            				pkg/disttask/framework/framework_test.go:35
	            				pkg/disttask/framework/framework_test.go:124
	            				pkg/util/wait_group_wrapper.go:157
	            				src/runtime/asm_amd64.s:1650
	Error:      	Not equal: 
	            	expected: "succeed"
	            	actual  : "failed"
	            	
	            	Diff:
	            	--- Expected
	            	+++ Actual
	            	@@ -1,2 +1,2 @@
	            	-(proto.TaskState) (len=7) "succeed"
	            	+(proto.TaskState) (len=6) "failed"
	            	 
	Test:       	TestFrameworkWithQuery
goroutine 33667 [running]:
testing.(*common).logDepth(0xc000e95a00, {0xc005bf2580, 0x284}, 0x3)
	GOROOT/src/testing/testing.go:1022 +0x4c5
testing.(*common).log(...)
	GOROOT/src/testing/testing.go:1004
testing.(*common).Errorf(0xc000e95a00, {0x24cd1de?, 0x53a5ee5?}, {0xc006987bc0?, 0x1c735c0?, 0x201?})
	GOROOT/src/testing/testing.go:1068 +0x54
github.com/stretchr/testify/assert.Fail({0x7f674c505c70, 0xc000e95a00}, {0xc000083ce0, 0xad}, {0x0, 0x0, 0x0})
	external/com_github_stretchr_testify/assert/assertions.go:333 +0x350
github.com/stretchr/testify/assert.Equal({0x7f674c505c70, 0xc000e95a00}, {0x1d866c0?, 0x2e7abb0}, {0x1d866c0?, 0xc006987ab0?}, {0x0, 0x0, 0x0})
	external/com_github_stretchr_testify/assert/assertions.go:417 +0x2bd
github.com/stretchr/testify/require.Equal({0x2e9a5a0, 0xc000e95a00}, {0x1d866c0, 0x2e7abb0}, {0x1d866c0, 0xc006987ab0}, {0x0, 0x0, 0x0})
	external/com_github_stretchr_testify/require/require.go:162 +0xbe
pkg/disttask/framework/framework_test_test.submitTaskAndCheckSuccess({0x2ebdfd0?, 0xc0010b60c0?}, 0x2?, {0x24cdb30?, 0x772ecfd2?}, 0x98c9dc0?, 0xc15da32a5cb3dfe6?)
	pkg/disttask/framework/framework_test.go:44 +0x89
pkg/disttask/framework/framework_test_test.submitTaskAndCheckSuccessForBasic(...)
	pkg/disttask/framework/framework_test.go:35
pkg/disttask/framework/framework_test_test.TestFrameworkWithQuery.func1()
	pkg/disttask/framework/framework_test.go:124 +0x125
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run.func1()
	pkg/util/wait_group_wrapper.go:157 +0x59
created by github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).Run in goroutine 84
	pkg/util/wait_group_wrapper.go:155 +0x7c
--
Coverage runner: Not collecting coverage for failed test. 

```

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -935,13 +935,13 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	// If running job meets error, we will save this error in job Error
 	// and retry later if the job is not cancelled.
 	schemaVer, runJobErr = w.runDDLJob(d, t, job)
-	defer d.unlockSchemaVersion(job.ID)
 
 	d.mu.RLock()
 	d.mu.hook.OnJobRunAfter(job)
 	d.mu.RUnlock()
 
 	if job.IsCancelled() {
+		defer d.unlockSchemaVersion(job.ID)
 		w.sess.Reset()
 		err = w.HandleJobDone(d, job, t)
 		return 0, err
@@ -963,11 +963,13 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	err = w.registerMDLInfo(job, schemaVer)
 	if err != nil {
 		w.sess.Rollback()
+		d.unlockSchemaVersion(job.ID)
 		return 0, err
 	}
 	err = w.updateDDLJob(job, runJobErr != nil)
 	if err = w.handleUpdateJobError(t, job, err); err != nil {
 		w.sess.Rollback()
+		d.unlockSchemaVersion(job.ID)
 		return 0, err
 	}
 	writeBinlog(d.binlogCli, txn, job)
@@ -976,9 +978,11 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	if !w.ddlCtx.isOwner() {
 		// This TiDB instance is not DDL owner anymore, it should not commit any transaction.
 		w.sess.Rollback()
+		d.unlockSchemaVersion(job.ID)
 		return 0, dbterror.ErrNotOwner
 	}
 	err = w.sess.Commit()
+	d.unlockSchemaVersion(job.ID)
 	if err != nil {
 		return 0, err
 	}

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -849,11 +849,6 @@ func (w *JobContext) setDDLLabelForDiagnosis(jobType model.ActionType) {
 }
 
 func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
-	if !w.ddlCtx.isOwner() && w.tp != localWorker {
-		// This TiDB instance is not DDL owner anymore, it should not commit any transaction.
-		w.sess.Rollback()
-		return dbterror.ErrNotOwner
-	}
 	err := w.finishDDLJob(t, job)
 	if err != nil {
 		w.sess.Rollback()
@@ -866,6 +861,16 @@ func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
 	}
 	CleanupDDLReorgHandles(job, w.sess)
 	asyncNotify(d.ddlJobDoneCh)
+	return nil
+}
+
+func (w *worker) checkOwnerBeforeCommit() error {
+	if !w.ddlCtx.isOwner() && w.tp != localWorker {
+		// Since this TiDB instance is not a DDL owner anymore,
+		// it should not commit any transaction.
+		w.sess.Rollback()
+		return dbterror.ErrNotOwner
+	}
 	return nil
 }
 
@@ -940,6 +945,11 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	d.mu.hook.OnJobRunAfter(job)
 	d.mu.RUnlock()
 
+	if err = w.checkOwnerBeforeCommit(); err != nil {
+		d.unlockSchemaVersion(job.ID)
+		return 0, err
+	}
+
 	if job.IsCancelled() {
 		defer d.unlockSchemaVersion(job.ID)
 		w.sess.Reset()
@@ -975,12 +985,6 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	writeBinlog(d.binlogCli, txn, job)
 	// reset the SQL digest to make topsql work right.
 	w.sess.GetSessionVars().StmtCtx.ResetSQLDigest(job.Query)
-	if !w.ddlCtx.isOwner() {
-		// This TiDB instance is not DDL owner anymore, it should not commit any transaction.
-		w.sess.Rollback()
-		d.unlockSchemaVersion(job.ID)
-		return 0, dbterror.ErrNotOwner
-	}
 	err = w.sess.Commit()
 	d.unlockSchemaVersion(job.ID)
 	if err != nil {

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -849,6 +849,11 @@ func (w *JobContext) setDDLLabelForDiagnosis(jobType model.ActionType) {
 }
 
 func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
+	if !w.ddlCtx.isOwner() && w.tp != localWorker {
+		// This TiDB instance is not DDL owner anymore, it should not commit any transaction.
+		w.sess.Rollback()
+		return dbterror.ErrNotOwner
+	}
 	err := w.finishDDLJob(t, job)
 	if err != nil {
 		w.sess.Rollback()

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -849,6 +849,9 @@ func (w *JobContext) setDDLLabelForDiagnosis(jobType model.ActionType) {
 }
 
 func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
+	if err := w.checkOwnerBeforeCommit(); err != nil {
+		return err
+	}
 	err := w.finishDDLJob(t, job)
 	if err != nil {
 		w.sess.Rollback()
@@ -935,18 +938,14 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	d.mu.hook.OnJobRunAfter(job)
 	d.mu.RUnlock()
 
-	if !w.ddlCtx.isOwner() {
-		// Since this TiDB instance is not a DDL owner anymore,
-		// it should not commit any transaction.
-		w.sess.Rollback()
-		d.unlockSchemaVersion(job.ID)
-		return 0, dbterror.ErrNotOwner
-	}
-
 	if job.IsCancelled() {
 		defer d.unlockSchemaVersion(job.ID)
 		w.sess.Reset()
 		err = w.HandleJobDone(d, job, t)
+		return 0, err
+	}
+
+	if err = w.checkOwnerBeforeCommit(); err != nil {
 		return 0, err
 	}
 
@@ -995,6 +994,16 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	}
 
 	return schemaVer, nil
+}
+
+func (w *worker) checkOwnerBeforeCommit() error {
+	if !w.ddlCtx.isOwner() && w.tp != localWorker {
+		// Since this TiDB instance is not a DDL owner anymore,
+		// it should not commit any transaction.
+		w.sess.Rollback()
+		return dbterror.ErrNotOwner
+	}
+	return nil
 }
 
 // HandleDDLJobV2 handles v2 ddl job.

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -864,16 +864,6 @@ func (w *worker) HandleJobDone(d *ddlCtx, job *model.Job, t *meta.Meta) error {
 	return nil
 }
 
-func (w *worker) checkOwnerBeforeCommit() error {
-	if !w.ddlCtx.isOwner() && w.tp != localWorker {
-		// Since this TiDB instance is not a DDL owner anymore,
-		// it should not commit any transaction.
-		w.sess.Rollback()
-		return dbterror.ErrNotOwner
-	}
-	return nil
-}
-
 func (w *worker) prepareTxn(job *model.Job) (kv.Transaction, error) {
 	err := w.sess.Begin()
 	if err != nil {
@@ -945,9 +935,12 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	d.mu.hook.OnJobRunAfter(job)
 	d.mu.RUnlock()
 
-	if err = w.checkOwnerBeforeCommit(); err != nil {
+	if !w.ddlCtx.isOwner() {
+		// Since this TiDB instance is not a DDL owner anymore,
+		// it should not commit any transaction.
+		w.sess.Rollback()
 		d.unlockSchemaVersion(job.ID)
-		return 0, err
+		return 0, dbterror.ErrNotOwner
 	}
 
 	if job.IsCancelled() {

--- a/pkg/ddl/ddl_worker.go
+++ b/pkg/ddl/ddl_worker.go
@@ -930,13 +930,13 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	// If running job meets error, we will save this error in job Error
 	// and retry later if the job is not cancelled.
 	schemaVer, runJobErr = w.runDDLJob(d, t, job)
+	defer d.unlockSchemaVersion(job.ID)
 
 	d.mu.RLock()
 	d.mu.hook.OnJobRunAfter(job)
 	d.mu.RUnlock()
 
 	if job.IsCancelled() {
-		defer d.unlockSchemaVersion(job.ID)
 		w.sess.Reset()
 		err = w.HandleJobDone(d, job, t)
 		return 0, err
@@ -958,20 +958,22 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 	err = w.registerMDLInfo(job, schemaVer)
 	if err != nil {
 		w.sess.Rollback()
-		d.unlockSchemaVersion(job.ID)
 		return 0, err
 	}
 	err = w.updateDDLJob(job, runJobErr != nil)
 	if err = w.handleUpdateJobError(t, job, err); err != nil {
 		w.sess.Rollback()
-		d.unlockSchemaVersion(job.ID)
 		return 0, err
 	}
 	writeBinlog(d.binlogCli, txn, job)
 	// reset the SQL digest to make topsql work right.
 	w.sess.GetSessionVars().StmtCtx.ResetSQLDigest(job.Query)
+	if !w.ddlCtx.isOwner() {
+		// This TiDB instance is not DDL owner anymore, it should not commit any transaction.
+		w.sess.Rollback()
+		return 0, dbterror.ErrNotOwner
+	}
 	err = w.sess.Commit()
-	d.unlockSchemaVersion(job.ID)
 	if err != nil {
 		return 0, err
 	}

